WDL/runtime/backend/singularity.py

```diff
@@ -49,11 +67,16 @@ def cli_exe(self) -> List[str]:
     def _pull_invocation(self, logger: logging.Logger, cleanup: ExitStack) -> Tuple[str, List[str]]:
         image, invocation = super()._pull_invocation(logger, cleanup)
         docker_uri = "docker://" + image
-
-        # The docker image layers are cached in SINGULARITY_CACHEDIR, so we don't need to keep the
-        # *.sif
-        pulldir = cleanup.enter_context(tempfile.TemporaryDirectory(prefix="miniwdl_sif_"))
-        return (docker_uri, self.cli_exe + ["pull", "--dir", pulldir, docker_uri])
+        if self.image_cache_dir is not None:
+            image_name = docker_uri.replace("/", "_").replace(":", "_")
+            image_path = os.path.join(self.image_cache_dir, image_name)
+            if not os.path.exists(image_path):
```

Comment:
The concurrency risks with this existence check are a bit tricky; I think it's OK but probably worth documenting:

* We're *assuming* there won't be other `miniwdl run` processes operating in this particular `image_cache_dir`. (if I understood correctly your reply to my earlier question)
* Our base class [already serializes pull operations](https://github.com/chanzuckerberg/miniwdl/blob/7a27f5ebda995b804b7af27b3656753c060601a0/WDL/runtime/backend/cli_subprocess.py#L184-L198) using a lock across all task runner threads in the current process.
* Therefore when `os.path.exists(image_path)`, that image file is either already complete or *will be* once the current thread acquires that lock.

A small risk is that a partially-written *.sif file was left behind by some previously-crashed runner. Perhaps there would be some use to pulling to a temporary filename and then atomically renaming it to `image_path` once complete. We can come back to that later since it's not convenient to do in the current code structure, and not a common/likely problem (and users can be advised to try wiping the cache directory).

Reply:
> We're assuming there won't be other miniwdl run processes operating in this particular image_cache_dir. (if I understood correctly your reply to my earlier question)

It is impossible to enforce a filelock. The linux kernel can enforce exclusive write access to a file. I can implement this on top, it does however complicate support on the other two major OSes.

> A small risk is that a partially-written *.sif file was left behind by some previously-crashed runner. Perhaps there would be some use to pulling to a temporary filename and then atomically renaming it to image_path once complete. We can come back to that later since it's not convenient to do in the current code structure, and not a common/likely problem (and users can be advised to try wiping the cache directory).

This is more common than you might think. 

1. Starting the workflow.
2. Remembering some option you forgot in the input JSON.
3. CTRL-C CTRL-C CTRL-C CTRL-C

This can corrupt a lot of caches very badly. Come to think about it, this was the primary reason SINGULARITY_CACHEDIR was so annoying. It had to be regularly wiped clean.

I will make the changes.